bftengine/src/bftengine/KeyExchangeManager.cpp

```diff
@@ -90,36 +117,33 @@ std::string KeyExchangeManager::onKeyExchange(const KeyExchangeMsg& kemsg,
     return "invalid epoch";
   }
   // reject key exchange message if generated seq_num is outside working window
-  if (sn < ((static_cast<uint64_t>(req_sn) / kWorkWindowSize) * kWorkWindowSize)) {
+  if (sn < ((req_sn / kWorkWindowSize) * kWorkWindowSize)) {
     LOG_WARN(KEY_EX_LOG, "Generated sequence number is outside working window, ignore..." << KVLOG(sn, req_sn));
     return "gen_seq_num_ouside_workingwindow";
   }
   if (publicKeys_.keyExists(kemsg.repID, sn)) return "ok";
-  publicKeys_.push(kemsg, sn);
-  if (kemsg.repID == repID_ && seq_candidate_map_.count(sn)) {  // initiated by me
-    private_keys_.key_data().generated = seq_candidate_map_[sn].generated;
-    seq_candidate_map_[sn].generated.clear();
-    candidate_private_keys_.generated.clear();
-    // erasing seqnum from the map
-    seq_candidate_map_.erase(sn);
-    ConcordAssert(private_keys_.key_data().generated.pub == kemsg.pubkey);
-    private_keys_.onKeyExchange(cid, sn);
-    for (auto e : registryToExchange_) e->onPrivateKeyExchange(private_keys_.key_data().keys[sn], kemsg.pubkey, sn);
-    metrics_->self_key_exchange_counter++;
-  } else {  // initiated by others
-    for (auto e : registryToExchange_) e->onPublicKeyExchange(kemsg.pubkey, kemsg.repID, sn);
-    metrics_->public_key_exchange_for_peer_counter++;
-  }
+
+  registerNewKeyPair(kemsg.repID, sn, kemsg.pubkey, cid);
+
   auto liveQuorumSize = ReplicaConfig::instance().waitForFullCommOnStartup ? clusterSize_ : quorumSize_;
   if (ReplicaConfig::instance().getkeyExchangeOnStart() && (publicKeys_.numOfExchangedReplicas() <= liveQuorumSize))
     LOG_INFO(KEY_EX_LOG,
              "Exchanged [" << publicKeys_.numOfExchangedReplicas() << "] out of [" << liveQuorumSize << "]"
-                           << KVLOG(kemsg.repID));
+                           << KVLOG(kemsg.repID, initial_exchange_, exchanged()));
   if (!initial_exchange_ && exchanged()) {
     initial_exchange_ = true;
     if (ReplicaConfig::instance().getkeyExchangeOnStart() &&
-        ReplicaConfig::instance().publishReplicasMasterKeyOnStartup)
+        ReplicaConfig::instance().publishReplicasMasterKeyOnStartup) {
       sendMainPublicKey();
+      sentKeyUpdateToExecution = true;
+    }
+  }
+
+  // The key was exchanged successfully, write a block containing the new key to the chain
```

Comment:
The preferred way to do it would be to have both functions in the same message (in case this feature is enabled).
I see the complications of doing so (because of multiple mechanisms (such as key exchange, reconfiguration), and so on).
The potential problem I can see with two different mechanisms is that we have a gap between the consensus exchange and the main key exchange.
For example. assuming the system is highly overloaded and the main key message is executed after X >>> 300 sequence numbers, will it cause any issues?
IMO, it is safe (for now), because:
1. For the replicas, the actual exchange happens already now, the mainKeyPublicKey message is meant to clients (as they fetch the key state using CRE)
2. Clients, however, don't care about consensus keys, so they shouldn't care about this gap.

Do you think it worth a comment here, only to say there might be an uncontrollable gap between the two executions?

Reply:
I think that the issue is the lack of atomicity in this operation, which affects both clients and the current replica state (key exchange reflected in reserved pages but not in block storage).
I mentioned this issue in the following lines of this comment.